src/Bounds.cpp

```diff
@@ -1039,11 +1039,29 @@ class Bounds : public IRVisitor {
                     if (op->is_intrinsic(Call::shift_left)) {
                         if (t.is_int() && t.bits() >= 32) {
                             // Overflow is UB
-                            if (a_interval.has_lower_bound() && b_interval.has_lower_bound() && can_prove(b_interval.min >= 0 && b_interval.min < t.bits())) {
+                            // The following two Exprs are to avoid warning
+                            // messages that abs of an unsigned type is a no-op.
+                            if (a_interval.has_lower_bound() &&
+                                b_interval.has_lower_bound() &&
+                                can_prove(b_interval.min >= 0 &&
+                                          b_interval.min < t.bits())) {
                                 interval.min = a_interval.min << b_interval.min;
+                            } else if (a_interval.has_lower_bound() &&
+                                       b_interval.has_lower_bound() &&
+                                       can_prove(b_interval.min < 0 &&
+                                                 cast<int>(b_interval.min) > -t.bits())) {
```

Comment:
Casting this to int seems possibly dangerous, if it is an unsigned int. Maybe adding to the condition: `... && !b_interval.min.type().is_uint() && ...` and then skipping the cast which is hopefully now unnecessary?

Reply:
I figured it would be fine since the one before is checking that it's < 0 and an unsigned min would fail there, but I agree that this is even safer.